File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/home/ModeFilterSheet.kt
package lk.mageride.passenger.home

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FilterChip
import androidx.compose.material3.FilterChipDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import lk.mageride.passenger.R
import lk.mageride.passenger.ui.component.MageRideCta
import lk.mageride.passenger.ui.component.SectionLabel
import lk.mageride.passenger.ui.theme.ControlTokens
import lk.mageride.passenger.ui.theme.MageRideTheme
import lk.mageride.shared.data.models.ServiceMode
import lk.mageride.shared.data.models.VehicleType

/**
 * SCR-PA-006 — the mode and type filter.
 *
 * An M3 `ModalBottomSheet` over the map, exactly as the wireframe draws it: three mode rows with
 * their A/B/C badges, a divider, then the type chips.
 *
 * **Every chip carries its own vehicle icon tinted with its MAP-03 colour**, which the wireframe
 * calls out in as many words — *"not a plain dot … so the chip matches the map marker colour"*. The
 * colour comes from the same `VehicleColors.Legend` the `SymbolLayer` tints with, so the two cannot
 * drift.
 *
 * **Apply closes the sheet and nothing else.** The filter is *"instant client-side"* (the wireframe
 * again) — every toggle has already redrawn the map underneath, so the CTA is a dismiss, not a
 * commit. Making it a commit would mean a passenger could not see what they were choosing.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun ModeFilterSheet(
    filter: MapFilter,
    onModeChanged: (ServiceMode, Boolean) -> Unit,
    onTypeChanged: (VehicleType, Boolean) -> Unit,
    onDismiss: () -> Unit,
) {
    ModalBottomSheet(onDismissRequest = onDismiss, sheetState = rememberModalBottomSheetState()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .padding(horizontal = MageRideTheme.spacing.md)
                .padding(bottom = MageRideTheme.spacing.lg),
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Text(
                text = stringResource(R.string.filter_title),
                style = MaterialTheme.typography.titleLarge,
                color = MaterialTheme.colorScheme.onSurface,
            )

            MapFilter.MODES.forEach { mode ->
                ModeRow(
                    mode = mode,
                    enabled = mode in filter.modes,
                    onChange = { onModeChanged(mode, it) },
                )
            }

            SectionLabel(
                text = stringResource(R.string.filter_vehicle_types),
                modifier = Modifier.padding(top = MageRideTheme.spacing.xs),
            )

            FlowRow(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
                verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xxs),
            ) {
                MapFilter.CHIP_TYPES.forEach { type ->
                    TypeChip(
                        type = type,
                        enabled = type in filter.types,
                        onChange = { onTypeChanged(type, it) },
                    )
                }
            }

            MageRideCta(
                label = stringResource(R.string.filter_apply),
                onClick = onDismiss,
                modifier = Modifier.padding(top = MageRideTheme.spacing.xs),
            )
        }
    }
}

/** One mode row — badge, name, switch. */
@Composable
private fun ModeRow(mode: ServiceMode, enabled: Boolean, onChange: (Boolean) -> Unit) {
    val modes = MageRideTheme.mode
    val colour = when (mode) {
        ServiceMode.A -> modes.modeA
        ServiceMode.B -> modes.modeB
        ServiceMode.C -> modes.modeC
    }

    Row(
        modifier = Modifier.fillMaxWidth(),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
    ) {
        Box(
            modifier = Modifier
                .size(ControlTokens.ListRowIcon)
                .background(colour, RoundedCornerShape(MageRideTheme.radius.sm)),
            contentAlignment = Alignment.Center,
        ) {
            Text(
                text = VehicleLabels.modeBadge(mode),
                style = MaterialTheme.typography.labelMedium,
                color = MaterialTheme.colorScheme.onPrimary,
                textAlign = TextAlign.Center,
            )
        }
        Text(
            text = stringResource(VehicleLabels.modeName(mode)),
            style = MaterialTheme.typography.bodyLarge,
            color = MaterialTheme.colorScheme.onSurface,
        )
        Box(modifier = Modifier.weight(1f))
        Switch(checked = enabled, onCheckedChange = onChange)
    }
}

/** One type chip — the wireframe's `chip` with its `vico` swatch. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun TypeChip(type: VehicleType, enabled: Boolean, onChange: (Boolean) -> Unit) {
    val tint = VehicleLabels.colour(MageRideTheme.vehicle, type)

    FilterChip(
        selected = enabled,
        onClick = { onChange(!enabled) },
        label = { Text(stringResource(VehicleLabels.name(type))) },
        leadingIcon = {
            // The swatch IS the marker colour — a disc of the legend tint with the type's own
            // glyph on it, so a chip and the pin it controls are recognisably the same thing.
            Box(
                modifier = Modifier
                    .size(ControlTokens.ListRowIcon)
                    .background(tint, CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    imageVector = VehicleLabels.icon(type),
                    contentDescription = null,
                    modifier = Modifier.size(ControlTokens.ChipIcon),
                    tint = MaterialTheme.colorScheme.onPrimary,
                )
            }
        },
        colors = FilterChipDefaults.filterChipColors(
            selectedContainerColor = MaterialTheme.colorScheme.primaryContainer,
            selectedLabelColor = MaterialTheme.colorScheme.onPrimaryContainer,
        ),
    )
}

File: apps/passenger-ios/PassengerApp/Home/ModeFilterSheet.swift
import SwiftUI
import UIKit

/// SCR-PI-006 — the mode and type filter.
///
/// A presented `.sheet` over the map, exactly as the cell draws it: a scrim, a grabber, *"Show on
/// map"*, three mode rows with their A/B/C badges, a `vehicle types` divider, the chip row, and
/// **Apply**. Its `Δ iOS` clause is *"`Toggle` + `Toggle(.button)` chips; `.impact(.light)`"*, and
/// this is all three.
///
/// **Every chip carries its own vehicle glyph tinted with its MAP-03 colour**, which the cell calls
/// out in as many words — *"each type shows its small vehicle icon tinted with the canonical AL-09
/// colour (not a plain dot)"*. The colour comes from the same ``VehicleToken`` the map's
/// `MLNSymbolStyleLayer` is tinted from, so the two cannot drift.
///
/// **Apply closes the sheet and nothing else.** The filter is *"instant"* (the cell again) — every
/// toggle has already redrawn the map underneath — so the CTA is a dismiss, not a commit. Making it
/// a commit would mean a passenger could not see what they were choosing.
struct ModeFilterSheet: View {

    let filter: MapFilter
    let onModeChanged: (ModeToken, Bool) -> Void
    let onTypeChanged: (VehicleToken, Bool) -> Void
    let onDismiss: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: MageRideSpacing.xs) {
            Text(key: "filter_title")
                .mageFont(.title)
                .foregroundStyle(MageRideColor.onSurface)
                .padding(.bottom, MageRideSpacing.xxs)

            GroupedList {
                ForEach(VehicleLabels.modeRows, id: \.self) { mode in
                    ModeRow(
                        mode: mode,
                        isOn: filter.modes.contains(mode),
                        // Key the collection and ask it whether this is the last row, rather than
                        // enumerating: Swift has no key paths into tuples, so `Array(_:enumerated())`
                        // cannot supply a `ForEach` id.
                        showsSeparator: mode != VehicleLabels.modeRows.last,
                        onChange: { onModeChanged(mode, $0) }
                    )
                }
            }

            LabelledDivider(key: "filter_vehicle_types")
                .padding(.top, MageRideSpacing.xxs)

            ChipGrid(types: VehicleLabels.chipTypes, selected: filter.types, onChange: onTypeChanged)

            Button(action: onDismiss) {
                Text(key: "filter_apply")
            }
            .buttonStyle(.mageCta)
            .padding(.top, MageRideSpacing.xs)
        }
        .padding(.horizontal, MageRideSpacing.md)
        .padding(.top, MageRideSpacing.md)
        .padding(.bottom, MageRideSpacing.lg)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(MageRideColor.surface)
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
    }
}

/// One mode row — the badge, the name, and iOS's green switch.
///
/// Drawn here rather than through ``GroupedRow``, and the reason is the badge: that control's
/// leading slot is a 28pt rounded square holding an **SF Symbol**, and the wireframe's `.badge` is a
/// small pill holding a **letter**. A row is cheaper than bending a shared control into a shape its
/// other four call sites do not want.
private struct ModeRow: View {

    let mode: ModeToken
    let isOn: Bool
    let showsSeparator: Bool
    let onChange: (Bool) -> Void

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: MageRideSpacing.sm) {
                Text(mode.badge)
                    .mageFont(.label)
                    .foregroundStyle(MageRideColor.onStatus)
                    .padding(.horizontal, MageRideSpacing.xs)
                    .padding(.vertical, MageRideSpacing.xxs / 2)
                    .background(mode.color, in: RoundedRectangle(cornerRadius: MageRideRadius.sm, style: .continuous))

                Text(key: mode.nameKey)
                    .mageFont(.body)
                    .foregroundStyle(MageRideColor.onSurface)

                Spacer(minLength: MageRideSpacing.xs)

                Toggle("", isOn: Binding(get: { isOn }, set: onChange))
                    .labelsHidden()
            }
            .padding(.horizontal, MageRideSpacing.sm)
            .frame(minHeight: MageRideControl.minimumTapTarget)

            if showsSeparator {
                Rectangle()
                    .fill(MageRideColor.surfaceVariant)
                    .frame(height: MageRideControl.hairline)
                    .padding(.leading, MageRideSpacing.sm)
            }
        }
        .accessibilityElement(children: .combine)
    }
}

/// The chip row. It wraps, because eight chips do not fit on one line at any Dynamic Type size.
///
/// A `LazyVGrid` with an `.adaptive` column rather than a hand-written `Layout`: the chips are a
/// fixed, short list, and an adaptive grid is what SwiftUI already ships for exactly this. (The
/// driver app needed a `Layout` because its rows hold labels of wildly different widths.)
private struct ChipGrid: View {

    let types: [VehicleToken]
    let selected: Set<VehicleToken>
    let onChange: (VehicleToken, Bool) -> Void

    var body: some View {
        LazyVGrid(
            columns: [
                GridItem(
                    .adaptive(minimum: MageRideControl.filterChipMinimum),
                    spacing: MageRideSpacing.xs,
                    alignment: .leading
                ),
            ],
            alignment: .leading,
            spacing: MageRideSpacing.xs
        ) {
            ForEach(types, id: \.self) { type in
                TypeChip(type: type, isOn: selected.contains(type)) { onChange(type, $0) }
            }
        }
    }
}

/// One type chip — the wireframe's `.chip` with its `.vico` swatch.
///
/// `Toggle(isOn:).toggleStyle(.button)` is the cell's own `Δ iOS` clause, and it is what gives the
/// chip its selected fill, its `.isSelected` accessibility trait and its announcement for free. A
/// hand-drawn button would have to earn all three.
private struct TypeChip: View {

    let type: VehicleToken
    let isOn: Bool
    let onChange: (Bool) -> Void

    var body: some View {
        Toggle(isOn: Binding(get: { isOn }, set: onChange)) {
            HStack(spacing: MageRideSpacing.xxs + 2) {
                // The swatch IS the marker colour — a disc of the legend tint with the type's own
                // glyph on it, so a chip and the pins it controls are recognisably the same thing.
                type.image
                    .font(.system(size: MageRideControl.chipIcon))
                    .foregroundStyle(MageRideColor.onStatus)
                    .frame(width: MageRideControl.chipSwatch, height: MageRideControl.chipSwatch)
                    .background(type.color, in: Circle())

                Text(key: type.nameKey)
                    .mageFont(.bodySmall)
                    .lineLimit(1)
            }
        }
        .toggleStyle(.button)
        .buttonStyle(.bordered)
        .tint(MageRideColor.primary)
        // The cell's own `.impact(.light)`. A chip that redraws every marker underneath it is worth
        // one; the mode switches already have the system's own.
        .onChange(of: isOn) { _ in UIImpactFeedbackGenerator(style: .light).impactOccurred() }
    }
}
